### ISSUE TITLE
Add dumbbell and slope charts to R-instat 

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#6523]
In our graphs it would be good to provide more of the graphs that we keep seeing.  We need to "make-it-easy" but we should also stick to some principles, so we are adding in a ggplot sort of way.  I suggest we can do this by adding a few more buttons to the existing specific dialogues.  One example is the lollipop charts.  

See [here](https://www.r-graph-gallery.com/301-custom-lollipop-chart.html#baseline) for many examples, with the corresponding R code when you click on a graph.

As explained here, a lollipop chart is basically a geom_segment and geom_point.  

Hence I suggest we have 2 buttons (for now) for the Describe > Specific > ScatterPlot.  The first is Points and is the current dialogue.  I also suggest we change the name from scatterplot to Point plot.  This means the same thing, but the name then is the same as the geom.  (That's what we do unless there is a good reason to be different.)

Then we add Segment as a second button.

We already do this with some plots that go together - like boxplot, which has jitter and violin.

In this case the second aspect is that we could do the points first and then add the segments, or vice versa.  I suggest (here) we only need to "Add Points" to the segment option.  We don't bother to "Add Segments" to the points option.  

The principle is that once we have an initial geom we can always use the Plot Options to add layers.  But there are many popular graphs that have multiple layers.  Here is a good example, where we want many options for segments under that button, and then we often want to add points also.  

@lilyclements has R code for lollipop plots from her MSc days, and so will look at this item first on the R side.  Then I hope @haward can include this in the (newly named) point Plot dialogue.

@dannyparsons has included lollipop in the Procurement menu.  I assume possibly as a special bar chart.  I prefer as here, partly because it is easy to include as double, and also cherry plots.  They seem to be an alternative to bubble plots, Perhaps we should also consider adding size of the points to the main point dialogue?

**Comment by rdstern** [comment]
I now have a further option.  [Here](https://alesandrab.wordpress.com/2014/10/15/how-to-create-a-double-lollipop-chart/) is an article on creating a double lollipop chart in Excel.  There are many steps and it is quite a process!

I suggest "double lollipops" are just another instance - in ggplot2 - of a segment layer plus a point layer.  
And the article listed above also mentions that they are like cherry charts - also line plus segment.  And that a cherry chart is an alternative to bubble charts - which is just a point plot - including the size (and possibly colour) of the points.

This confirms my enthusiasm for adding Segment as another button to the scatterplot dialogue and also of renaming the plot as Point Plot to relate to geom_point.

![image](https://user-images.githubusercontent.com/11226469/121703248-8b0cab80-caca-11eb-8a4a-2558abe4558a.png)

I also recognise these charts above as simple examples of double lollipops!

Lots of reasons for doing this!

**Comment by lilyclements** [comment]
I've had a little look at this. For segment plots you should give a starting and ending value for both `x` and `y` options.
* For a horizontal segment, the start and end `y` values should be the same.
* For a vertical segment, the start and end `x` values should be the same. 
* You can also have diagonal options if the start and end values are different for both `x` and `y`.

e.g.
![image](https://user-images.githubusercontent.com/21180424/122652457-dfe99b00-d136-11eb-9f66-d30bbae441eb.png)

I have a few ideas for the design, but before I continue I just want to check: @rdstern do we want the user to have the option to create a diagonal segments?
If that is the case, then we can have four receivers: "Starting x-value", "Ending x-value", "Starting y-value", "Ending y-value"
Or do we want a default vertical segments and a checkbox to instead have horizontal segments?

Additionally, the value for `x`, `xend`, `y`, `yend` can be a factor or numeric variable, or just a fixed numeric value given in a nud.
If `xend` is a factor, `x` must also be a factor (note that this is not the case if `x` is a factor).
Hence if the variable in the "Starting X" receiver is numeric, then the "Ending X" receiver should only accept numeric variables (or can be a specified fixed value).
If the variable in the "Ending X" receiver is a factor, then the "Starting X" receiver should only accept factor variables.
This is similarly the case for `yend` and `y`.

**Comment by rdstern** [comment]
Of course I'd like all!

Simple lollipop charts would need just a single point, because they start at an axis.  Some examples are [here](https://www.data-to-viz.com/graph/lollipop.html).  Then, as you say, the double, needs the 2 points.  Here is a [recent example](http://www.rebeccabarter.com/blog/2018-05-29_getting_fancy_ggplot2/) with diagonal.  It uses what we will make simple, namely geom_segment, plus geom_point, plus geom_text.  Quite a lot of code for our 3 layers!

![image](https://user-images.githubusercontent.com/11226469/122794535-434d0780-d2b4-11eb-86ac-bc2656135d1e.png)

I am assuming we will also have (simple) lollipop charts on our 2 variable (and 3-variable) menu.  The specific charts are moving the user a step towards flexibility, with understanding - so the addition of the layers is explicit in these dialogues.  With up to 3 layers, namely segment, plus points, plus labels/texts.  

So you already get a lot of flexibility with our specific charts.  2 variable and 3 variable is less.  General plotting is more, i.e. building the layers one-by-one.  So, if the diagonal lines is making the dialogue too messy, then we just allow vertical or horizontal and leave the diagonal to the general system.

**Comment by lilyclements** [comment]
Okay great. How about something like this?

![image](https://user-images.githubusercontent.com/21180424/123119588-93a2a180-d43b-11eb-824b-d28369e2e8a5.png)

1. All the current options on the dialog go into the "Points" button, and we have a new option called "Segment"
2. There are four receivers - I've called these A, B, C, and D for now since their label changes slightly depending on which radio buttons are selected in 3. and 4.
3. There is a group box with three radio buttons: Simple, Double, and Diagonal. Simple is the default.
4. There is another group box titled "Orientation" with two radio buttons: Vertical and Horizontal. Vertical is the default.
5. There is a checkbox titled "Add Points".

Should I move the group boxes to be before the selector/receivers?

The R code alters slightly depending on which radio buttons are selected, additionally different receivers are visible so I'll go through the options below.
In the following R codes I have included after a comment the additional code if "Add Points" is selected (the `geom_point` code).

**Simple - Vertical**
2 receivers and 1 input box with labels
* X Variable (A)
* Y Variable (C)
* Starting Value

```
ggplot(df) +
  geom_segment(data = df, aes(x = x, xend = x, y=<<value in "starting value" input box>>, yend=y)) +
  # if "Add Points" is selected also run:
  geom_point(data = df, aes(x = x, y = y))
```

**Simple - Horizontal**
Same layout as with Simple - Vertical

```
ggplot(df) +
  geom_segment(data = df, aes(x = <<value in "starting value" input box>>, xend = x, y = y, yend = y)) +
  # if "Add Points" is selected also run:
  geom_point(data = df, aes(x = x, y = y))
```

**Double - Vertical**
3 receivers with labels
* X Variable (A)
* Starting Y Variable (C)
* Ending Y Variable (D)

```
ggplot(df) +
  geom_segment(data = df, aes(x = x, xend = x, y = start_y, yend = end_y)) +
  # if "Add Points" is selected also run:
  geom_point(data = df, aes(x = x, y = starty)) +
  geom_point(data = df, aes(x = x, y = endy), shape = 1)
```

**Double - Horizontal**
3 receivers with labels
* Starting X Variable (A)
* Ending X Variable (B)
* Y Variable (C)

```
ggplot(df) +
  geom_segment(data = df, aes(x = start_x, xend = end_x, y = y, yend = y)) +
  # if "Add Points" is selected also run:
  geom_point(data = df, aes(x = start_x, y = y)) +
  geom_point(data = df, aes(x = end_x, y = y), shape = 1)
```

**Diagonal**
3 receivers with labels
* Starting X Variable (A)
* Ending X Variable (B)
* Starting Y Variable (C)
* Ending Y Variable (D)

```
ggplot(df) +
  geom_segment(data = df, aes(x = start_x, xend = end_x, y = start_y, yend = end_y)) +
  # if "Add Points" is selected also run:
  geom_point(data = df, aes(x = start_x, y = start_y)) +
  geom_point(data = df, aes(x = end_x, y = end_y), shape = 1)
```

**Comment by HawardKetoyoMsatsi** [comment]
@lilyclements This is very clear and well explained I am happy to do the front-end implementation ASAP.

**Comment by rdstern** [comment]
@HawardKetoyoMsatsi this sounds good.  In @lilyclements diagram I suggest the Points option will have a few consequences - not just a single checkbox, but perhaps size and colour included.
Then @shadrackkibet is working on labelling for the bar chart dialogue.  I suggest the same (geom_text) layer will also be added to this dialogue.

**Comment by rdstern** [comment]
I am wondering if I have led you astray in the work above.  I have now found (again) the ggalt package, and this includes geom_lollipop and geom_dumbell!  Could @lilyclements and @HawardKetoyoMsatsi please look at them.  Should we include these geoms instead of (effectively) constructing them ourselves?

If so, then I am slightly inclined to include them on the bar chart dialogue?

**Comment by rdstern** [comment]
I have now looked a bit further at these (new to me) geoms etc.

Let's start with Lollipop charts.  I took data from the hh dataframe from the Pakistan MICS survey, and got the percentage using Urdu from each district.  Here is my data frame:

And here is the dialogue:

![image](https://user-images.githubusercontent.com/11226469/125188924-a1269c80-e22d-11eb-9937-4d6dcb1c6b37.png)

with the resulting bar chart:

![image](https://user-images.githubusercontent.com/11226469/125189034-33c73b80-e22e-11eb-9304-696387fb9941.png)

The code is as follows:

```
# Code generated by the dialog, Bar and Pie Chart

hh_by_HH7 <- data_book$get_data_frame(data_name="hh_by_HH7")
last_graph <- ggplot2::ggplot(data=hh_by_HH7, mapping=ggplot2::aes(x=HH7, y=percent_urdu)) + ggplot2::geom_bar(position="dodge", stat="identity") + theme_grey() + ggplot2::coord_flip()
data_book$add_graph(graph_name="last_graph", graph=last_graph, data_name="hh_by_HH7")
data_book$get_graphs(data_name="hh_by_HH7", graph_name="last_graph")
rm(list=c("last_graph", "hh_by_HH7"))

```
Now, suppose we ass ggalt and code geom_lollipop.  Then just changing geom_bar to geom_lollipop i.e

```
hh_by_HH7 <- data_book$get_data_frame(data_name="hh_by_HH7")
last_graph <- ggplot2::ggplot(data=hh_by_HH7, mapping=ggplot2::aes(reorder(x=HH7,X=percent_urdu), y=percent_urdu)) + ggalt::geom_lollipop() + theme_grey() + ggplot2::coord_flip()
data_book$add_graph(graph_name="last_graph", graph=last_graph, data_name="hh_by_HH7")
data_book$get_graphs(data_name="hh_by_HH7", graph_name="last_graph")
rm(list=c("last_graph", "hh_by_HH7"))
```
This seems a "bargain, so we could do this!?
Here is the resulting lollipop graph.
![image](https://user-images.githubusercontent.com/11226469/125190700-5b220680-e236-11eb-8194-a1ff933e9777.png)

You will notice I have also added the sorting - which would be equally useful for the bar charts.  It just adds the reorder function to the R code.  So Order could be a new checkbox.  If checked, then there could be another checkbox, with ascending as the label.  If checked it simply does minus the sorting variable.

I couldn't quite see how to do this with the frequency option and then ..count.., which I think is now better written as stat(count)?

If putting this option in as a possible checkbox, then also consider allowing the colour and size of the point to be specified on the main dialogue.  And the special option button would change to Options Lollipop

 

.

**Comment by lilyclements** [comment]
For `geom_dumbell`:

1. "The dumbbell geom is used to create dumbbell charts. Dumbbell dot plots--dot plots with two or more series of data--are an alternative to the clustered bar chart or slope graph."
Can you explain a bit about what a dot plot with more than two series of data would look like? Does this mean there would be more than two points on the dumbell plot (e.g., instead of a "start" and "end", a "start", "middle", and "end"? If so, I can see a way to do this, but it's not very nice at all! It still only allows a "start" and "end" variable only, and the middle is added by layering over it)
2. A `y`, `x`, and `xend` parameter must always be given, and these parameters can be either a variable or a fixed value. Do we want to offer this in form of three receivers with a checkbox next to each one to offer to switch the receiver to a fixed value?


For `newggslopegraph`:
1. It does work with facets. Using the survey data:

```
newggslopegraph(survey,
                variety,
                yield,
                fertgrp,
                Title = NULL,
                SubTitle = NULL,
                Caption = NULL) +
  theme_gray() +
  #labs(title = "overwrite", subtitle = "hello") +
  facet_wrap(vars(village))
```

![image](https://user-images.githubusercontent.com/21180424/125480649-2bffbaa4-2859-47f5-a33e-21b99fcd8d25.png)

2. The `Times` parameter only takes characters, factors, or ordered factors. The `Measurement` parameter only takes numeric variables

3. Should the `Title`, `SubTitle`, and `Caption` parameters should be `NULL` by default in R-Instat? In the R function, their default values are `No title given` (and the like) which isn't consistent with other plots? The `NULL` labels can be overwritten by the `labs` layer in the sub-dialog so the user can add titles/etc to this plot by the same method as with other plots.
Similarly, should we have `+ theme_gray()` for consistency?

4. A lot of the options in the `newggslopegraph` function concern titles/subtitles/captions (size, placement, etc) which can already be done by different geoms already coded in the sub-dialog so I don't feel the need for these options? (Additionally, they seem very temperamental for me, but perhaps I'm having issues with the laptop)

5. There are also options with labels. I'm not sure how the labels code is coming along in R-Instat, however, I assume the labelling options from this function can be overwritten by label geoms?
Are the labelling options being added in as a new layer/tab in a sub-dialog?
Would we want to offer labelling through the R code in the `newggslopegraph` function, or through a geom (or neither)?

6. Other parameter options, such as line width/colour etc, look useful. These parameters and their values are:
* `LineThickness = 1`
* `LineColor = "ByGroup"`
* `ReverseYAxis = FALSE`
* `ReverseXAxis = FALSE`
* `RemoveMissing = TRUE`

**Comment by rdstern** [comment]
@lilyclements I reply to some of your comments.

"The dumbbell geom is used to create dumbbell charts. Dumbbell dot plots--dot plots with two or more series of data--are an alternative to the clustered bar chart or slope graph."

I was merely noting that the dumbbell geom mentioned it was for "2 or more series", and I couldn't easily see how.  I am quite happy if it is just (for us) for 2 series - we don't have to do everything.

I get the impression you are ok with adding the geom_dumbbell instead of constructing it ourselves (with segment + point).  If so, then the next task is to add geom_lollipop and geom_dumbbell to our list?

I wonder about your view of newggslopegraph?  As you say it isn't our usual way and hence many of the options overlap with aspects we do differently, i.e. we would be using the function fairly minimally and then adding our usual aspects the ordinary way.  I like the idea of a slope graph, and his results look reasonable, but there is nothing else in his package that we are likely to need.  So, should we do it differently?

**Comment by HawardKetoyoMsatsi** [comment]
@rdstern  , having the geom_lolipop and geom_dumbel is a good idea compared to building the plots from segments and points as we did. I will be happy to close PR #6586 where I added the lollipop graphs that had been suggested by @lilyclements  in this issue were implemented and quickly add these two geoms (geom_lolipop and geom_dumbel) to PR #6605 where I added some geom parameters that had been left out for then to be merged as soon as possible. The charts can then be handled in  #6526 as suggested.

**Comment by HawardKetoyoMsatsi** [comment]
@rdstern  I have added geom_lolipop and geom_dumbel in PR #6605
In geom_dimbbell, documentation `yend` is given as a mandatory aesthetic but in running the function it is not mandatory so have including it under optional aesthetics

**Comment by lilyclements** [comment]
>  I couldn't easily see how. I am quite happy if it is just (for us) for 2 series - we don't have to do everything.

@rdstern I think I've managed to work this one out - if we give two factor variables for the `x` and `xend` values then we can have multiple series. However, I'm really not sure what the use is of this for interpretation!

```
ggplot(mtcars, aes(x=gear, xend = carb, y=cyl))+
  geom_dumbbell(size = 1, colour_x = "#5b8124", colour_xend = "#bad744") +
  labs(x="gear to carb")
```

![image](https://user-images.githubusercontent.com/21180424/125814912-5ae52408-15c2-4d5b-8c6d-b168af3ec769.png)

> I get the impression you are ok with adding the geom_dumbbell instead of constructing it ourselves (with segment + point). If so, then the next task is to add geom_lollipop and geom_dumbbell to our list?

@rdstern yes, I think adding the code using the `geom_dumbell` and `geom_lollipop` functions is a good way forward.

> I wonder about your view of newggslopegraph? As you say it isn't our usual way and hence many of the options overlap with aspects we do differently, i.e. we would be using the function fairly minimally and then adding our usual aspects the ordinary way. I like the idea of a slope graph, and his results look reasonable, but there is nothing else in his package that we are likely to need. So, should we do it differently?

I really like the look of the function for `newggslopegraph`. It seems very flexible, and labels/etc can be added still using the R-Instat dialogs so it is still compatible with everything else in R-Instat.
The main difference between this function and the R-code for a slope graph that I suggested previously in this issue is that this function requires only one variable for the x-axis (a factor). I assume this is more what we want in R-Instat (e.g. say we wanted to compare rain by year for each station then we can just use the "year" variable as the x variable).
I think we should keep `newggslopegraph`, however if you disagree then, looking at the function code, it is very trivial for us to replicate their R code and offer a slope graph by constructing through geoms.

**Comment by rdstern** [comment]
@lilyclements that's great.  Where to put the new graphs has now fallen into place too.
Looking at Chapter 7 of the [Data Visualisation with R ](https://rkabacoff.github.io/datavis/Time.html#TimeSeries)book it has 4 types of graph in its time-series chapter, as follows:

![image](https://user-images.githubusercontent.com/11226469/125916416-025ce904-219e-4a31-ad59-a20636d37947.png)

So, I propose we add (probably) 3 buttons at the top.  They could be `Line, Dumbbell, Slope, Area `so they correspond with the chapter headings.  

a) Dumbbell - I hope easy to implement now that @HawardKetoyoMsatsi has added the geom
b) Slope:  This could perhaps still include a bit more @lilyclements investigation.  The default is for it to use the newggslopeplot function from the CRAN package (CGPfunctions).  An alternative is that there is also a function called ggslopegraph in an R package by Leeper, though not on CRAN.  Perhaps that version could be used by copying over the function?
The special sub-dialogue would be different.  I wonder if it would take long to adapt the code to produce our own geom_slope as an adaptation of one or other function?
c) Perhaps we also implement Area plots so they tie completely with the chapter?  We already have geom_area.

**Comment by rdstern** [comment]
On the design of the slope graph I think we do need to use the newggslopegraph function.  It assumes the data in the "right shape" for us.  
I am still not sure we need this whole package?  (CGPfunctions) partly as it installs many other packages.  Is it easy to just take the function - with attribution, and possibly then simplification?

The dialogue is I think largely our single/multiple receiver.  (That's for the Y).  Then x variable as usual.  If we use multiple, then those variables become the x-variable (that's different to our usual line plot.  Then there is a Group, or Factor - not both and it isn't optional.  We get one slope line for each level of the factor.  If single, then it is usually a factor of a few time points as the x-variable.

**Comment by lilyclements** [comment]
I agree that the `newggslopegraph` function seems the best option. It should be very easy to copy over the function and simplify it (and give full attribution!).

**Comment by lilyclements** [comment]
If we simplify the `newggslopegraph` function we can remove some of the options, such as the title/caption/etc. Should I keep in their default labelling options?
It is quite important in the graphic I think, and then the user can edit these in a sub.

I think @shadrackkibet is looking at labels. This plot uses the `ggrepel` package which gives more flexible options with labelling (#6549).
If I were to keep the labels when copying over the function, we should discuss which labelling to use. Otherwise the `geom_text` function in the `ggplot` package just pastes labels on top of the `geom_text_repel` function in the `ggrepel` package, rather than replaces.

**Comment by lilyclements** [comment]
@rdstern shall I write a new function simplifying ‘newggslopegraph’ to fit what we want, or should we use ‘newggslopegraph’?

Having it as a new function will mean we don’t have to install the relatively large package that this function is in, and additionally would mean less issues conflicting if the user wanted to add labels to the plot (since this function uses different labels to us, from memory unless that’s changed)

The only issue I can see with rewriting the function lays in the time aspect as I will have to learn how to write a ‘+ geom’ function.
If we do want to do a new function simply if the ‘ggnewslopegraph’ we could write it as a function for now, and then later on write it as it’s own geom.

**Comment by rdstern** [comment]
Yes, please add the function

**Comment by lilyclements** [comment]
Here is the function - I've split it into two for now:
* function to run the code
* theme for the function.

The code to run it on the dialog is

`newggslopegraph_amended(data = <Data Frame>, x = <X Variable>, y = <Y Variable>, colour = <Colour/By variable>) + newggslopegraph_theme()`

e.g.
```
data(newcancer)
newggslopegraph_amended(newcancer, Year, Survival, Type) + newggslopegraph_theme()
```

I've made some very minor amendments (edited the error messages to be more consistent with R-Instat errors, used embraces for the programming instead of the `enquo` system since the braces system is newer, changed parameter names, etc)

When there is more time I can look into creating it as a geom so that it is even more consistent with R-Instat.

@HawardKetoyoMsatsi can you add this function into the stand_alone_functions.R file

@rdstern which dialog did we want this to be a part of? 

[newggslopegraph function - separating for RI.zip](https://github.com/africanmathsinitiative/R-Instat/files/7120683/newggslopegraph.function.-.separating.for.RI.zip)

Variables in the function:
* data, x, y, color, data_label = NULL, 
          y_text_size = 3, 
          line_thickness = 1, line_color = "ByGroup", 
          data_text_size = 2.5, data_text_color = "black", data_label_padding = 0.05, 
          data_label_line_size = 0, data_label_fill_color = "white", 
          reverse_x_axis = FALSE, 
          remove_missing = TRUE

**Comment by rdstern** [comment]
@lilyclements great, many thanks.

The dumbbell and slope are both in the Line plot dialogue, because line plots (and also these are often for time-series data.

Just to check, that the dumbbell option didn't need anything from you, @lilyclements , because it is a geom already - like geom_lolliop is is in the ggalt package.  Your work was for the slope function?

**Comment by lilyclements** [comment]
Yes. This work was for the slope function just amending ‘newggslopegraph’. This isn’t as a geom at the moment, just as a function.
Hopefully can further separate it into a geom in the future.

**Comment by HawardKetoyoMsatsi** [comment]
@lilyclements @rdstern the dumbel plot is ready however I cannot open a PR since some other people are working on the line plot dialog.  PRs #6724 and #6708 should be merged first before we can open a PR fixing the Line plot dialog. @rdstern could you please look ate these previously opened PRs to pave way for us to open other PRs for dumbell and newsggslope plots.
